### PR TITLE
8311993: Test serviceability/sa/UniqueVtableTest.java failed: duplicate vtables detected

### PR DESCRIPTION
**Pull request by alexmenkov**
On Windows SA agent gets a class vtable from symbols, exported from jvm.dll (it exports symbols like "??_7" + type + "@@6B@").
But symbol lookup function first requests WinDbg about the symbol.
Sometimes WinDbg routine IDebugSymbols::GetOffsetByName() returns offset for both class and class pointer types. Returned offsets correspond to symbols like "jvm!class_name::`vftable'".
The behavior is intermittent, I was not able to find what is the reason.
The fix adds workaround for the case - if GetOffsetByName succeeded, we check if corresponding symbol contains requested one.
So it returns expected offset for non-vtable symbols like "MaxJNILocalCapacity" (GetOffsetByName returns offset for "jvm!MaxJNILocalCapacity"), but returns 0 for vtlb lookup.

Additionally added check for results of IDebugSymbols::SetImagePath/SetSymbolPath

Testing: tier1,tier2,hs-tier5-svc

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8311993](https://bugs.openjdk.org/browse/JDK-8311993): Test serviceability/sa/UniqueVtableTest.java failed: duplicate vtables detected (**Bug** - P3)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) Review applies to [e42280af](https://git.openjdk.org/jdk/pull/20684/files/e42280afdfe899f29c9fd00d933761a2143ed5ac)
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20684/head:pull/20684` \
`$ git checkout pull/20684`

Update a local copy of the PR: \
`$ git checkout pull/20684` \
`$ git pull https://git.openjdk.org/jdk.git pull/20684/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20684`

View PR using the GUI difftool: \
`$ git pr show -t 20684`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20684.diff">https://git.openjdk.org/jdk/pull/20684.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20684#issuecomment-2305530184)

### CODE REVIEW

File: src/jdk.hotspot.agent/windows/native/libsaproc/sawindbg.cpp [hunk 1]
```diff
@@ -852,10 +857,22 @@ JNIEXPORT jlong JNICALL Java_sun_jvm_hotspot_debugger_windbg_WindbgDebuggerLocal
   if (ptrIDebugSymbols->GetOffsetByName(name, &offset) != S_OK) {
     return (jlong) 0;
   }
+
+  // See JDK-8311993: WinDbg intermittently returns offset of "module!class::`vftable'" symbol
+  // when requested for decorated "class" or "class*" (i.e. "??_7class@@6B@"/"??_7class*@@6B@").
+  // As a workaround check if returned symbol contains requested symbol.
+  ULONG64 disp = 0L;
+  char buf[512];
```
Comment (dholmes-ora) [inline src/jdk.hotspot.agent/windows/native/libsaproc/sawindbg.cpp:865]: s/512/SYMBOL_BUFSIZE/ ?

**Review by dholmes-ora — APPROVED** [review]
Not an expert on this code by any means, but this seems a reasonable way to tackle the problem.

Thanks.

**Comment by plummercj** [comment]
Why do we rely on GetOffsetByName() when we can already lookup directly from the dll, and this is in fact the fallback already in place when GetOffsetByName() fails?

File: src/jdk.hotspot.agent/windows/native/libsaproc/sawindbg.cpp (hunk 1 above)
Comment (alexmenkov) [reply src/jdk.hotspot.agent/windows/native/libsaproc/sawindbg.cpp]: Thank you, fixed

**Comment by alexmenkov** [comment]
> I found this:
> 
> ```
>     // should we parse DLL symbol table in Java code or use
>     // Windbg's native lookup facility? By default, we use
>     // native lookup so that we can take advantage of '.pdb'
>     // files, if available.
>     useNativeLookup = true;
>     String str = System.getProperty("sun.jvm.hotspot.debugger.windbg.disableNativeLookup");
>     if (str != null) {
>       useNativeLookup = false;
>     }
> ```
> 
> I'm not sure what is meant by "take advantage of '.pbp' files". Is it perhaps a more reliable or complete database of symbols, or perhaps it is faster? In any case, I'd be interested in seeing if all our tests still pass when useNativeLookup is false.

As far as I understand .pdb files contain symbol information.
I run tier1,tier2,hs-tier5-svc with "-Dsun.jvm.hotspot.debugger.windbg.disableNativeLookup=yes" on windows-x64 and windows-x64-debug
Got 110 failures; all failures on debug build, most of them with NPE:

```
Error: java.lang.NullPointerException: Cannot invoke "sun.jvm.hotspot.debugger.Address.getJBooleanAt(long)" because "address" is null
java.lang.NullPointerException: Cannot invoke "sun.jvm.hotspot.debugger.Address.getJBooleanAt(long)" because "address" is null
	at jdk.hotspot.agent/sun.jvm.hotspot.runtime.VM.isSharingEnabled(VM.java:940)
	at jdk.hotspot.agent/sun.jvm.hotspot.types.basic.BasicTypeDataBase.findDynamicTypeForAddress(BasicTypeDataBase.java:240)
        ...
```

**Comment by plummercj** [comment]
```
  public boolean isSharingEnabled() {
    if (sharingEnabled == null) {
        Address address = VM.getVM().getDebugger().lookup(null, "UseSharedSpaces");
        if (address == null && getOS().equals("win32")) {
            // On Win32 symbols are prefixed with the dll name. So look for
            // UseSharedSpaces as a symbol in jvm.dll.
            address = VM.getVM().getDebugger().lookup(null, "jvm!UseSharedSpaces");
        }
        sharingEnabled = address.getJBooleanAt(0);
    }
    return sharingEnabled.booleanValue();
  }
```

I'm not sure why this is failing. Based on the existing code and comments, it seems at some point SA worked without relying on the windbg native symbol support. Code like isSharingEnabled() probably got added afterwards and was never tested with it disabled.

**Comment by sspitsyn** [comment]
> I'm not sure why this is failing. Based on the existing code and comments, it seems at some point SA worked without relying on the windbg native symbol support. Code like isSharingEnabled() probably got added afterwards and was never tested with it disabled.

It seems the `isSharingEnabled()` has a bug and does not always work with the option`Dsun.jvm.hotspot.debugger.windbg.disableNativeLookup=yes"`.
So, there are two choices:
 - use this work around from Alex (with a potential overhead because of extra calls to `GetOffsetByName()`)
 - fix the `isSharingEnabled()` related bug and use `GetOffsetByName()` instead of native lookup

I wonder if the call to`GetOffsetByName()` can be done with some additional check, so it is not always called.

File: src/jdk.hotspot.agent/windows/native/libsaproc/sawindbg.cpp [hunk 2]
```diff
@@ -852,10 +857,22 @@ JNIEXPORT jlong JNICALL Java_sun_jvm_hotspot_debugger_windbg_WindbgDebuggerLocal
...
     return (jlong) 0;
   }
+
+  // See JDK-8311993: WinDbg intermittently returns offset of "module!class::`vftable'" symbol
+  // when requested for decorated "class" or "class*" (i.e. "??_7class@@6B@"/"??_7class*@@6B@").
+  // As a workaround check if returned symbol contains requested symbol.
+  ULONG64 disp = 0L;
+  char buf[SYMBOL_BUFSIZE];
+  memset(buf, 0, sizeof(buf));
+  if (ptrIDebugSymbols->GetNameByOffset(offset, buf, sizeof(buf), 0, &disp) ==  S_OK) {
```
Comment (sspitsyn) [inline src/jdk.hotspot.agent/windows/native/libsaproc/sawindbg.cpp:867]: Nit: Extra space before `S_OK`.

**Comment by plummercj** [comment]
Alex's changes neither introduce nor fix the isSharingEnabled() bug. It was just discovered out of an investigation to better understand how SA is doing symbol lookups on Windows. Basically we are trying to understand why it tries both windbg and dll lookups. Currently it won't work with just dll lookups, as the isSharingEnabled() failures show, but if that is the case how can dll lookups be a reliable fallback when windbg lookups fail.

**Comment by alexmenkov** [comment]
> ```
>   public boolean isSharingEnabled() {
>     if (sharingEnabled == null) {
>         Address address = VM.getVM().getDebugger().lookup(null, "UseSharedSpaces");
>         if (address == null && getOS().equals("win32")) {
>             // On Win32 symbols are prefixed with the dll name. So look for
>             // UseSharedSpaces as a symbol in jvm.dll.
>             address = VM.getVM().getDebugger().lookup(null, "jvm!UseSharedSpaces");
>         }
>         sharingEnabled = address.getJBooleanAt(0);
>     }
>     return sharingEnabled.booleanValue();
>   }
> ```
> 
> I'm not sure why this is failing. Based on the existing code and comments, it seems at some point SA worked without relying on the windbg native symbol support. Code like isSharingEnabled() probably got added afterwards and was never tested with it disabled.

"UseSharedSpaces" is exported from jvm.dll, but the issue here on Windows SymbolLookup searches for decorated symbols (i.e. "??_7UseSharedSpaces@@6B@") and the symbol is exported undecorated.
From globalDefinition.hpp (note extern "C"):

```
extern "C" {
// Make sure UseSharedSpaces is accessible to the serviceability agent.
extern JNIEXPORT jboolean UseSharedSpaces;
}
```

**Comment by plummercj** [comment]
What would happen if it was not extern "C". Would the windbg lookup succeed in that case?

**Comment by plummercj** [comment]
Actually I'm a bit confused on this. The extern "C" is making it so the dll lookup fails, but is it helping the windbg lookup to succeed? If yes, why did it ever get added if it was always first succeeding with the dll lookup.

**Comment by plummercj** [comment]
> Update: usually constants are available for SA by using VMStruct, I'm not sure why "UseSharedSpaces" is exported this way

Probably the author just didn't know about the VMStruct approach.

**Comment by alexmenkov** [comment]
> What would happen if it was not extern "C". Would the windbg lookup succeed in that case?

without extern "C" the variable is exported as "?UseSharedSpaces@@3EA". So it wouldn't work with DLL lookup neither windbg.

**Comment by alexmenkov** [comment]
Anyway failures with disableNativeLookup and UseSharedSpaces is different issue.
I can file JBS issue, but to me disableNativeLookup looks like development property.

**Comment by plummercj** [comment]
Where do you see runtime flags in vmstructs?

**Comment by alexmenkov** [comment]
> Where do you see runtime flags in vmstructs?

I do not see :) I thought about constants, but they are different

**Comment by plummercj** [comment]
I can't find any other place where we directly lookup a global symbol like this. I did find this other reference to the lookup() API:

```
  public String findSymbol(String symbol) {
    Address addr = lookup(null, symbol);
    if (addr == null && getOS().equals("win32")) {
      // On win32 symbols are prefixed with the dll name. Do the user
      // a favor and see if this is a symbol in jvm.dll or java.dll.
      addr = lookup(null, "jvm!" + symbol);
      if (addr == null) {
        addr = lookup(null, "java!" + symbol);
      }
    }
    ...
```

So it seems isSharingEnabled() could be using this instead, although that is not solving any of the problems we are seeing with non-native lookups. clhsdb findsym uses this findSymbol() API.

```
        new Command("findsym", "findsym name", false) {
            public void doit(Tokens t) {
                if (t.countTokens() != 1) {
                    usage();
                } else {
                    String result = VM.getVM().getDebugger().findSymbol(t.nextToken());
                    out.println(result == null ? "Symbol not found" : result);
                }
            }
        },
```

We have one test that does a findsym of MaxJNILocalCapacity (ClhsdbFindPC.java). I'm guessing this would fail if native lookups were disabled. Note findsym was added less than 3 years ago.

Another thing I noticed is that isSharingEnabled() used to be the following:

  public boolean isSharingEnabled() {
    if (sharingEnabled == null) {
      Flag flag = getCommandLineFlag("UseSharedSpaces");
      sharingEnabled = (flag == null)? Boolean.FALSE :
          (flag.getBool()? Boolean.TRUE: Boolean.FALSE);
    }
    return sharingEnabled.booleanValue();
  }

So it used to use getCommandLineFlag(). It was modified to use lookup() as part of https://bugs.openjdk.org/browse/JDK-8277481, which changed UseSharedSpaces from a command line flag to just a regular global. This change was done 3 years ago. getCommandLineFlag() seems to get the flag from VMStructs, which references the following:

JVMFlag* JVMFlag::flags = flagTable;
size_t JVMFlag::numFlags = (sizeof(flagTable) / sizeof(JVMFlag));

So there is no dll or windbg symbol lookup here, just access to the VMStructs database.

It seems then that theoretically the dll lookup should never be needed, but due to the bug this PR is fixing, it is needed as a fallback when windbg lookup fails to do the right thing. I wonder if there are also other types of symbols we lookup that would only be found with the dll lookup.

**Review by plummercj — APPROVED** [review]
Changes look good.

**Comment by alexmenkov** [comment]
> We have one test that does a findsym of MaxJNILocalCapacity (ClhsdbFindPC.java). I'm guessing this would fail if native lookups were disabled. Note findsym was added less than 3 years ago.

The test fails on isSharingEnabled, but I suppose it will fail to find MaxJNILocalCapacity too

> It seems then that theoretically the dll lookup should never be needed, but due to the bug this PR is fixing, it is needed as a fallback when windbg lookup fails to do the right thing. I wonder if there are also other types of symbols we lookup that would only be found with the dll lookup.

As far as I understand dll lookup is needed to get vtables for classes (and I suppose it's the only case).
I don't think it's "right thing" to return offset of "jvm!class_name::`vftable'" for decorated symbols (and especially for pointers).
Maybe vtable lookup can be implemented by searching for "class_name::`vftable'", then dll lookup (and all this exported symbols) can be dropped

**Review by kevinjwalls — APPROVED** [review]
I thought it odd that we lookup symbols including the *, e.g.
Duplicate vtable: 0x00007ffd233633d8:
  - CompiledMethod (extends CodeBlob)
  - CompiledMethod* (extends null)

And iterating over  agent.getTypeDataBase().getTypes(); we see e.g. jbyte, jbyte*, jbyte** which looks odd and unnecessary.

We know what a pointer is, so not needing to lookup up "jbyte*" might be good also.
But that might be more change than we want to do right now, so this double check looks good.

**Pull request closed without merging**